voter-dapp/src/RetrieveRewards.js

```diff
@@ -0,0 +1,201 @@
+
+import { useTableStyles } from "./Styles.js";
+import { MAX_UINT_VAL } from "./common/Constants.js";
+
+// Max integer that can be safely stored in a vanilla js int.
+const MAX_SAFE_INT = 2147483647;
+
+function getOrCreateObj(containingObj, field) {
+  if (!containingObj[field]) {
+    containingObj[field] = {};
+  }
+
+  return containingObj[field];
+}
+
+function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceResolvedEvents, votingAccount) {
+  const { drizzle, useCacheSend } = drizzleReactHooks.useDrizzle();
+  const { web3 } = drizzle;
+
+  const { send, status } = useCacheSend("Voting", "retrieveRewards");
+
+  if (retrievedRewardsEvents === undefined || revealedVoteEvents === undefined || priceResolvedEvents === undefined) {
+    // Requests haven't been completed.
+    return { ready: false, status };
+  } else {
+    // Put events into a simple mapping from identifier|time -> state
+    const state = {};
+
+    // Function to get the voteState object for an identifier and time.
+    const getVoteState = (identifierIn, timeIn) => {
+      const identifier = web3.utils.hexToUtf8(identifierIn);
+      const time = timeIn.toString();
+
+      return getOrCreateObj(state, `${identifier}|${time}`);
+    };
+
+    // Each of the following loops adds the relevant portions of the event state to the corresponding voteState data
+    // structure.
+    for (const event of retrievedRewardsEvents) {
+      const voteState = getVoteState(event.returnValues.identifier, event.returnValues.time);
+
+      voteState.retrievedRewards = true;
+    }
+
+    for (const event of priceResolvedEvents) {
+      const voteState = getVoteState(event.returnValues.identifier, event.returnValues.time);
+
+      voteState.priceResolutionRound = event.returnValues.resolutionRoundId.toString();
+    }
+
+    // Since this loop is the last one, we can use the state to determine if this identifer, time, roundId combo
+    // is eligible for a reward claim. We track the oldestUnclaimedRound to determine which round the transaction
+    // should query (since only one can be chosen).
+    let oldestUnclaimedRound = MAX_SAFE_INT;
+    for (const event of revealedVoteEvents) {
+      const voteState = getVoteState(event.returnValues.identifier, event.returnValues.time);
+      const revealRound = event.returnValues.roundId.toString();
+
+      if (!voteState.retrievedRewards && voteState.priceResolutionRound === revealRound) {
+        // Reveal happened in the same round as the price resolution: definitely a retrievable reward.
+        oldestUnclaimedRound = Math.min(oldestUnclaimedRound, revealRound);
+        voteState.didReveal = true;
+      }
+    }
+
+    if (oldestUnclaimedRound === MAX_SAFE_INT) {
+      // No unclaimed rounds were found.
+      return { ready: true, status };
+    }
+
+    // Extract identifiers and times from the round we picked.
+    const toRetrieve = [];
+    const maxBatchRetrievals = 100;
```

Comment:
We'll want to move this constant to `common/Constants.js` like the limits for batchCommits and batchReveals, see the pending PR for this change #847

Reply:
Agreed. Will move in a follow up, since the other PR is still pending.